src/classes/RD_RecurringDonations.cls

```diff
@@ -795,277 +725,467 @@ public class RD_RecurringDonations {
+
+            if (needsRecalc) {
+                if(r.npe03__Open_Ended_Status__c == closedlabel
+                        && oldRecurringDonations.get(r.id).npe03__Open_Ended_Status__c != r.npe03__Open_Ended_Status__c
                     ){
-                    reevaluateopps.add(r.id);
+                    // closing an open ended donation
+                    newlyClosedDonations.add(r);
+                } else if (r.npe03__Open_Ended_Status__c == openlabel) {
+                    // open-ended
+                    openEndedRDsToRecalculate.put(r.Id, r);
+                } else {
+                    // fixed-length
+                    fixedLengthRDsToRecalculate.put(r.Id, r);
                 }
-                else if (r.npe03__Installment_Period__c != oldRD.npe03__Installment_Period__c || r.npe03__Contact__c != oldRD.npe03__Contact__c){
-                    reinsertopps.add(r);
-                }                                        
             }
-        }      
-        
-        if (!newlycloseddonations.isEmpty()){
-            runRecurringDonationClosedActions(newlycloseddonations, dmlWrapper);
         }
-        if (!reevaluateopps.isEmpty()){
-            if (reevaluateopps.size() == 1 || system.isFuture() || system.isBatch())
-                updateExistingOpps(reevaluateopps, dmlWrapper);
-            else
-                updateExistingOppsFuture(reevaluateopps);
+        
+        if (!newlyClosedDonations.isEmpty()){
+            runRecurringDonationClosedActions(newlyClosedDonations, dmlWrapper);
         }
-        //needs a total reinsert, not an update
-        if (!reinsertopps.isEmpty()){
-            //clear existing open opps and recreate
-            delete [select id from Opportunity where npe03__Recurring_Donation__c IN :reinsertopps and isClosed = false];
-            if (reinsertopps.size() == 1 || system.isFuture() || system.isBatch()){
-                insertOppsOnRecurringDonationInsert(reinsertopps);                  
-            }
-            else{
-                set<id> rdset = new set<id>();
-                for (npe03__Recurring_Donation__c r : reinsertopps)
-                    rdset.add(r.id);
-                insertOppsOnRecurringDonationInsertFuture(rdset);   
+
+        List<Opportunity> oppsFromOpenRDs = new List<Opportunity>();
+        List<Opportunity> oppsFromFixedLengthRDs = new List<Opportunity>();
+        List<Opportunity> allOppsToSort = [SELECT Id
+                , Amount
+                , Name
+                , CampaignId
+                , AccountID
+                , CloseDate
+                , StageName
+                , isClosed
+                , isWon
+                , Recurring_Donation_Installment_Number__c
+                , npe03__Recurring_Donation__c
+        FROM Opportunity
+        WHERE npe03__Recurring_Donation__c IN :openEndedRDsToRecalculate.keySet()
+        OR npe03__Recurring_Donation__c IN :fixedLengthRDsToRecalculate.keySet()
```

Comment:
You could rewrite the filter as follows so that the query only retrieves the records needed versus retrieving all the records for those RD's. I'm thinking about RD's that might have more than 50K attached Opportunities here, so even though adding the filters on isClosed/isWon are on non-indexed fields, I suspect reducing the number of queried records will help with overall performance.

`WHERE (npe03__Recurring_Donation__c IN :openEndedRDsToRecalculate.keySet() AND IsClosed = False)
OR npe03__Recurring_Donation__c IN :fixedLengthRDsToRecalculate.keySet()`

If this should included ClosedLost opps for openEndedRDsToRecalculate then the query would be:
`WHERE (npe03__Recurring_Donation__c IN :openEndedRDsToRecalculate.keySet() AND (IsClosed = False OR (IsClosed = True AND IsWon=False))
OR npe03__Recurring_Donation__c IN :fixedLengthRDsToRecalculate.keySet()`